homeassistant/components/sensor/fitbit.py

```diff
@@ -0,0 +1,311 @@
+        config_file = config_from_file(config_path, default_config)
+
+    import fitbit
+
+    access_token = config_file.get('access_token')
+    refresh_token = config_file.get('refresh_token')
+    if None not in (access_token, refresh_token):
+        authd_client = fitbit.Fitbit(config.get('client_id'),
+                                     config.get('client_secret'),
+                                     access_token=access_token,
+                                     refresh_token=refresh_token)
+
+        if int(time.time()) - config_file.get('last_saved_at', 0) > 3600:
+            authd_client.client.refresh_token()
+
+        authd_client.system = authd_client.user_profile_get()['user']['locale']
+
+        dev = []
+        for resource in config.get('monitored_resources',
+                                   FITBIT_DEFAULT_RESOURCE_LIST):
+            dev.append(FitbitSensor(authd_client, config_path, resource))
+        add_devices(dev)
+
+    else:
+        oauth = fitbit.api.FitbitOauth2Client(config.get('client_id'),
+                                              config.get('client_secret'))
+        redirect_uri = "http://127.0.0.1:8123/auth/fitbit/callback"
+
+        def _start_fitbit_auth(handler, path_match, data):
+            """Start Fitbit OAuth2 flow."""
+            url, _ = oauth.authorize_token_url(redirect_uri=redirect_uri,
+                                               scope=["activity", "heartrate",
+                                                      "nutrition", "profile",
+                                                      "settings", "sleep",
+                                                      "weight"])
+            handler.send_response(301)
+            handler.send_header('Location', url)
+            handler.end_headers()
+
+        def _finish_fitbit_auth(handler, path_match, data):
+            """Finish Fitbit OAuth2 flow."""
+            success_html = b"""<html><head><title>Fitbit Auth</title></head>
+            <body><h1>Fitbit has been successfully authorized! You can close
+            this window now!</h1></body></html>"""
+            handler.send_response(HTTP_OK)
+            handler.send_header("Content-type", "text/html")
+            handler.end_headers()
+            handler.wfile.write(success_html)
+            from oauthlib.oauth2.rfc6749.errors import MismatchingStateError
+            from oauthlib.oauth2.rfc6749.errors import MissingTokenError
+            if data.get('code') is not None:
+                try:
+                    oauth.fetch_access_token(data.get('code'), redirect_uri)
+                except MissingTokenError as error:
+                    _LOGGER.error("Missing token: %s", error)
+                except MismatchingStateError as error:
+                    _LOGGER.error("Mismatched state, CSRF error: %s", error)
+            else:
+                _LOGGER.error("Unknown error when authing")
```

Comment:
Also, should we continue the rest of this method when an error occurs or should we `return`?